### PR TITLE
[semver:minor] skip building image when tags already exist in ecr repo

### PR DESCRIPTION
**Pull request by ricoli**
### Checklist

- [x] All new jobs, commands, executors, parameters have descriptions
- [x] Examples have been added for any significant new features
- [x] README has been updated, if necessary

### Motivation, issues

When using a tagging strategy based on the git commit hash it is unnecessary to spend time rebuilding an image that has been already pushed to the ECR repository. This is particularly relevant when wanting to revert to a previous deployment in production, saving precious time that would have been otherwise spent building the same image that already exists in ECR.

### Description

* I've tried my best to keep things backwards compatible, hopefully succeeded at it?
* To avoid repetition I've moved the aws-cli install and configuration into the ecr-login step, as that's when it's needed.

### CODE REVIEW

**Comment by gmemstr** [comment]
@ricoli Mind resolving the merge conflicts before I go ahead and give this a formal review? :)

**Comment by ricoli** [comment]
> @ricoli Mind resolving the merge conflicts before I go ahead and give this a formal review? :)

sure thing, done now

File: src/commands/build-image.yml [hunk 1]
```diff
@@ -39,20 +39,56 @@ parameters:
...
-          -f <<parameters.path>>/<<parameters.dockerfile>> \
-          $docker_tag_args \
-          <<parameters.path>>
+        if [ "<<parameters.skip-build-when-tags-exist-in-ecr>>" = "false" ] || [ "<<parameters.skip-build-when-tags-exist-in-ecr>>" = "true" -a "$docker_tags_exist_in_ecr" = "false" ]; then
+          docker build \
+            <<#parameters.extra-build-args>><<parameters.extra-build-args>><</parameters.extra-build-args>> \
+            -f <<parameters.path>>/<<parameters.dockerfile>> \
+            $docker_tag_args \
+            <<parameters.path>>
+        fi
```
Comment (gmemstr) [inline src/commands/build-image.yml:91]: Can you confirm this logic has been tested locally? I'll be trying to confirm this myself but don't have a testing ground yet.

File: src/commands/ecr-login.yml [hunk 2]
```diff
@@ -23,7 +23,31 @@ parameters:
...
+    default: AWS_SECRET_ACCESS_KEY
+    description: >
+      AWS secret key for IAM role. Set this to the name of
+      the environment variable you will set to hold this
+      value, i.e. AWS_SECRET_ACCESS_KEY.
+
 steps:
+  - aws-cli/install
+
+  - aws-cli/configure:
```
[the remark this answers is not in the document]
Comment (ricoli) [reply src/commands/ecr-login.yml:45]: oh right, _that's_ what the conflict was about :D

File: src/commands/build-image.yml (hunk 1 above)
Comment (ricoli) [reply src/commands/build-image.yml:91]: I thought the tests in the circleci config would be running all the time? That being said, I didn't see any feedback coming back to the PR

Comment (ricoli) [reply src/commands/build-image.yml]: hi @gmemstr I've seen that you triggered the job in circleci - could you please retrigger https://app.circleci.com/pipelines/github/CircleCI-Public/aws-ecr-orb/276/workflows/ef4cfb8c-40b6-40db-bfe2-12c47f5fcb10 as I've pushed out a fix for the failure?

Comment (ricoli) [reply src/commands/build-image.yml]: and I had another issue, with the regex this time... any chance you could retrigger it please?

Comment (ricoli) [reply src/commands/build-image.yml]: sorry to be a pain but I've now pushed another commit to reorder the create repo step - any chance the correct workflow could be automatically run on every push? :)

Comment (ricoli) [reply src/commands/build-image.yml]: sorry, couple more things fixed :)

Comment (ricoli) [reply src/commands/build-image.yml]: and hopefully last one - tests need to be sequential as otherwise they delete each other's ecr repo

Comment (ricoli) [reply src/commands/build-image.yml]: finally - all green now :)

File: README.md [hunk 3]
```diff
@@ -75,6 +75,10 @@ workflows:
 
           # The amount of time to allow the docker build command to run before timing out (default is `10m`)
           no-output-timeout: 15m
+
+          # Set to true if you don't want to build the image if it already exists in the ECR repo, for example when
+          # you are tagging with the git commit hash. Specially useful for faster code reverts.
+          skip-build-when-tags-exist-in-ecr: false
```
Comment (gmemstr) [inline README.md:81]: A bit of a long parameter name, maybe could be shortened to something like `skip-when-tags-exist`?

Comment (ricoli) [reply README.md]: done

**Pull request merged**